agent/config/builder.go

```diff
@@ -797,6 +798,26 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
 		return RuntimeConfig{}, fmt.Errorf("serf_wan_allowed_cidrs: %s", err)
 	}
 
+	// Handle Deprecated UI config fields
+	if c.UI != nil {
+		b.warn("The 'ui' field is deprecated. Use the 'ui_config.enabled' field instead.")
+		if c.UIConfig.Enabled == nil {
+			c.UIConfig.Enabled = c.UI
+		}
+	}
```

Comment:
I think it would be easier to trace how this value is set (and more consistent with other fields) by removing this from here and changing line 1709 below to:

```go
	b.boolValWithDefault(c.UIConfig.Enabled, b.boolVal(c.UI))
```

The warning can be anywhere, it could stay here, but might also be good in `uiConfigVal`.

I think the same is true for the two settings below.

Reply:
I originally was going to do it that way but ended up not because:

 1. most other `somethingVal` methods only accept the sub-struct from config not the whole thing. In general the pattern is for those to be "happy path" setters/converters from one struct type to another. We'd have to pass the whole Config into the `uiConfigVal` method which is a little different from the prevailing pattern here.
 2. In general other logic which has to move across multiple fields in different parts of the `Config` are handled as special cases here - see the deprecation of acl_datacenter and passing of the value
 3. I can see what you mean about the setting of values all being in one place with your suggestion. This way does have the benefit that all the code for the deprecated fields is in one place and can potentially be removed all as one later. If I were trying to find this i might do so by grepping for the deprecated warning and having the actual defaulting in the same place seems useful rather than having this deprecation handled in multiple places.
 4. The precedent for use of the `WithDefault` values is all within the Builder method directly that has access to the full config - most are in this "prelude" special cases section already with a few that were just mechanical movements (acl_* to acl.*) being in the assignments directly.